crates/uv/src/commands/tool/install.rs

```diff
@@ -47,14 +47,37 @@ pub(crate) async fn install(
     if preview.is_disabled() {
         warn_user_once!("`uv tool install` is experimental and may change without warning.");
     }
-    let from = from.unwrap_or(name.clone());
+
+    let from = if let Some(from) = from {
+        let from_requirement = Requirement::<VerbatimParsedUrl>::from_str(&from)?;
+        // Check if the user provided more than just a name positionally or if that name conflicts with `--from`
+        if from_requirement.name.to_string() != package {
+            // Determine if its an entirely different package or a conflicting specification
+            let package_requirement = Requirement::<VerbatimParsedUrl>::from_str(&package)?;
+            if from_requirement.name != package_requirement.name {
+                bail!(
+                    "Package name `{}` provided with `--from` does not match install request `{}`",
+                    from_requirement.name,
+                    package
+                );
+            } else {
+                bail!(
+                    "Package requirement `{}` provided with `--from` conflicts with install request `{}`",
+                    from,
+                    package
+                );
+            }
```

Comment:
I presume there are some utilities around for checking if two requirement specifications are compatible with each other?